### PR TITLE
Automated cherry pick of #121624: use context for lazy evaluation.

### PR DESCRIPTION
**Pull request by liangyuanpeng**
Cherry pick of #121624 on release-1.29.

#121624: use context for lazy evaluation.

For details on the cherry pick process, see the [cherry pick requests](https://git.k8s.io/community/contributors/devel/sig-release/cherry-picks.md) page.

```release-note

```

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
cc the author of PR  @jiahuif

**Review by jiahuif — APPROVED** [review]
/release-note-none
/lgtm
Thank you.

**Comment by cici37** [comment]
Note: this fix also needs to be backported to 1.28

**Review by xmudrii — APPROVED** [review]
For RelEng:
/lgtm
/approve

**Comment by jiahuif** [comment]
Can we have the backport to 1.28 approved too? https://github.com/kubernetes/kubernetes/pull/122516

**Comment by xmudrii** [comment]
@jiahuif We can only cherry-pick-approve it once it has lgtm and approved labels. We'll be happy to approve it once that's sorted out

**Pull request merged**
